src/backend/distributed/transaction/transaction_management.c

```diff
@@ -803,3 +826,40 @@ TriggerMetadataSyncOnCommit(void)
 {
 	MetadataSyncOnCommit = true;
 }
+
+
+/*
+ * Function raises an exception, if the current backend started a coordinated
+ * transaction and got a PREPARE event to become a participant in a 2PC
+ * transaction coordinated by another node.
+ */
+static void
+EnsurePrepareTransactionIsAllowed(void)
+{
+	if (!InCoordinatedTransaction())
+	{
+		/* If the backend has not started a coordinated transaction */
+		return;
+	}
+
+	if (GetInForceDelegatedFuncExecution())
+	{
+		/*
+		 * If a distributed forceDelegation function is executed in the
+		 * current coordinated transaction, allow restricted 2PC
+		 */
+		return;
```

Comment:
I'm not sure why we have this check. The check below should be sufficient for all cases where we execute the forced function remotely, and if the forced function call goes through local execution (and other parts of the transaction do remote execution) then we should not allow PREPARE TRANSACTION.

Reply:
Removed it.